modules/ipp/scanner.go

```diff
@@ -126,86 +172,394 @@ func (scanner *Scanner) GetPort() uint {
+	cupsBody := getPrintersRequest(version.Major, version.Minor)
+	cupsReq, err := http.NewRequest("POST", scan.url, cupsBody)
 	if err != nil {
-		return 0, 0, zgrab2.DetectScanError(err)
+		return zgrab2.NewScanError(zgrab2.SCAN_UNKNOWN_ERROR, err)
+	}
+	cupsReq.Header.Set("Accept", "*/*")
+	cupsReq.Header.Set("Content-Type", ContentType)
+	cupsResp, err := scan.client.Do(cupsReq)
+
+	if versionNotSupported(cupsResp, scanner) {
+		// TODO: Make this step down a version number
+		return nil
+	}
+	scan.results.CUPSResponse = cupsResp
+	if cupsResp != nil && cupsResp.Body != nil {
+		defer cupsResp.Body.Close()
+	}
+	// Store data into BodyText and BodySHA256 of cupsResp
+	storeBody(cupsResp, scanner)
+
+	bodyBytes := []byte(cupsResp.BodyText)
+	cupsVersions, _ := readAttributeFromBody(CupsVersion, &bodyBytes)
+	if len(cupsVersions) > 0 {
+		// Report this cupsVersion b/c it's more detailed than the one found in Server header
+		scan.results.CUPSVersion = strings.Split(scan.results.CUPSVersion, "/")[0] + "/" + string(cupsVersions[0])
+	}
+	// TODO: Determine whether ipp-versions must be in increasing order, if not, sort them
+	// TODO: Overwrite ipp_version with highest version listed
+	ippVersions, _ := readAttributeFromBody(VersionsSupported, &bodyBytes)
+	if len(ippVersions) > 0 {
+		highestVersion := string(ippVersions[len(ippVersions)-1])
+		components := strings.Split(highestVersion, ".")
+
+		output, err := strconv.Atoi(components[0])
+		if err != nil {
+			//handle
+		}
+		major := int8(output)
+		output, err = strconv.Atoi(components[1])
+		if err != nil {
+			//handle
+		}
+		minor := int8(output)
+		// Only compare new values to previous if we have previous values
+		comparable := scan.results.MajorVersion != nil && scan.results.MinorVersion != nil
+		if !comparable || (comparable && major >= *scan.results.MajorVersion && minor > *scan.results.MinorVersion) {
+			scan.results.MajorVersion = &major
+			scan.results.MinorVersion = &minor
+		}
 	}
+	uris, _ := readAttributeFromBody(PrinterURISupported, &bodyBytes)
+	if len(uris) > 0 {
+		scan.results.PrinterURI = string(uris[0])
+	}
+	return nil
+}
+
+func (scanner *Scanner) Grab(scan *scan, target *zgrab2.ScanTarget, version *version) *zgrab2.ScanError {
+	// Send get-printer-attributes request to the host, preferably a print server
+	body := getPrinterAttributesRequest(version.Major, version.Minor, scan.url)
+	request, err := http.NewRequest("POST", scan.url, body)
+	if err != nil {
+		return zgrab2.NewScanError(zgrab2.SCAN_UNKNOWN_ERROR, err)
+	}
+	request.Header.Set("Accept", "*/*")
+	request.Header.Set("Content-Type", ContentType)
+	resp, err := scan.client.Do(request)
+	if versionNotSupported(resp, scanner) {
+		return zgrab2.NewScanError(zgrab2.SCAN_APPLICATION_ERROR, ErrVersionNotSupported)
+	}
+	//Store response regardless of error in request, because we may have gotten something back
+	scan.results.Response = resp
 	if resp != nil && resp.Body != nil {
 		defer resp.Body.Close()
```

Comment:
This should probably go right after you get `resp` back -- in e.g. line 349 it could be leaked (if it weren't already reading / closing it)